src/qat/model/hardware_base.py

```diff
@@ -1,3 +1,68 @@
+from __future__ import annotations
+
+from typing import Annotated, TypeVar, get_args
+
+import numpy as np
+from annotated_types import Predicate
+from pydantic import NonNegativeInt, RootModel
+
 # This base file is used to implement classes/methods common to all hardware.
 
-QubitId = int
+QubitId = NonNegativeInt
+
+
+def validate_calibratable_positive_float(v: CalibratablePositiveFloat):
+    if not np.isnan(v) and v < 0.0:
+        raise ValueError(f"Given value {v} must be >=0.")
+    return v
+
+
+CalibratablePositiveFloat = Annotated[
+    float,
+    Predicate(validate_calibratable_positive_float),
+    validate_calibratable_positive_float,
+]
+
+
+def validate_calibratable_unit_interval(v: CalibratableUnitInterval):
+    if not np.isnan(v):
+        if v < 0.0 or v > 1.0:
+            raise ValueError("Given value {v} must be in the interval [0, 1].")
+    return v
+
+
+CalibratableUnitInterval = Annotated[
+    float,
+    Predicate(validate_calibratable_unit_interval),
+    validate_calibratable_unit_interval,
+]
+
+K = TypeVar("K")
+V = TypeVar("V", CalibratablePositiveFloat, CalibratableUnitInterval)
+
+
+class ValidatedDict(RootModel[dict[K, V]]):
```

Comment:
Sadly, pydantic does not validate its mutable fields (such as dictionaries) after instantiation. See a similar discussion [here](https://stackoverflow.com/questions/75396856/pydantic-does-not-validate-the-key-values-of-dict-fields). This is why we need to give these mutable fields an extra layer of validation. This is currently a prototype which works for `physical_connectivity_quality` but should be general enough to be applied for all dictionaries within the hardware model.